File: code/CompositeExpressions/images/ImageToHex.py
from functools import reduce
from operator import mul
from PIL import Image
import sys

binToHex = {
    "0000": "0",
    "0001": "1",
    "0010": "2",
    "0011": "3",
    "0100": "4",
    "0101": "5",
    "0110": "6",
    "0111": "7",
    "1000": "8",
    "1001": "9",
    "1010": "a",
    "1011": "b",
    "1100": "c",
    "1101": "d",
    "1110": "e",
    "1111": "f",
}

# Convert binary value to hex value
def toHex(binary):
    return "0x" + binToHex[binary[:4]] + binToHex[binary[4:]]

# Convert 1D array to 2D array
def convert(lst, shape):
    # Stop is shape is 1D
    if len(shape) == 1:
        return lst
    
    n = reduce(mul, shape[1:])
    return [convert(lst[i * n : (i + 1) * n], shape[1:]) for i in range(len(lst) // n)]

# Check for image file
if len(sys.argv) == 2:
    print("Usage: python ImageToHex.py <img path>")

# Open image using PIL
path = sys.argv[1]
img = Image.open(path).convert("L")
width, height = img.size

# Convert pixels to black/white 2D array
pixels = []
for pixel in img.getdata():
    pixels.append(0 if pixel < 255 else 1)
pixels = convert(pixels, (width, height))

# Turn image into custom chars
chop = 8
tiles = {}
for i in range(0, width, chop):
    for j in range(0, height, chop):
        tile = []
        for k in range(chop):
            row = ""
            for l in range(chop):
                row += str(pixels[i + k][j + l])

            tile.append(row)

        tiles[f"From ({i},{j}) to ({i+chop},{j+chop}): "] = tile

# Print font values as string
vals = []
for tile, data in tiles.items():
    print(tile, end="")

    tileVals = []
    for row in data:
        val = toHex(row)
        vals.append(val)
        tileVals.append(val)

    print(", ".join(tileVals))

print("\nRaw Hex Data: ", ", ".join(vals))

File: code/ImageToHex/ImageToHex/ImageToHex.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <sstream>
#include <vector>
#include <map>

using namespace cv;

template <typename Iter>
std::string join(Iter begin, Iter end, std::string const& sep);
std::string toHex(std::string binary);

int main(int argc, char** argv) {
	if (argc != 2) {
		std::string progName = argv[0];
		std::cout << "Usage: " << progName.substr(progName.find_last_of("\\/") + 1) << " <img path>\n";
		return -1;
	}

	std::string path = argv[1];
	Mat img = imread(path, 0);
	const int width = img.rows, height = img.cols;

	std::vector<std::vector<int>> pixels(width, std::vector<int>(height));
	for (int i = 0; i < width; i++) {
		for (int j = 0; j < height; j++) {
			pixels[i][j] = img.at<uchar>(i, j) < 255 ? 0 : 1;
		}
	}

	const int chop = 8;
	std::map<std::string, std::vector<std::string>> tiles;
	for (int i = 0; i < width; i += chop) {
		for (int j = 0; j < height; j += chop) {
			std::vector<std::string> tile;
			for (int k = 0; k < chop; k++) {
				std::string row;
				for (int l = 0; l < chop; l++) {
					row += std::to_string(pixels[i + k][j + l]);
				}
				tile.push_back(row);
			}
			
			std::stringstream ss;
			ss << "From (" << i << ", " << j << ") to (" << i + chop << ", " << j + chop << "): ";
			tiles[ss.str()] = tile;
		}
	}

	std::vector<std::string> allVals;
	for (auto tile : tiles) {
		std::cout << tile.first;
		std::vector<std::string> vals;
		for (std::string row : tile.second) {
			std::string val = toHex(row);
			vals.push_back(val);
			allVals.push_back(val);
		}

		std::cout << join(vals.begin(), vals.end(), ", ") << "\n";
	}

	std::cout << "\nRaw Hex Data: " << join(allVals.begin(), allVals.end(), ", ") << "\n";
}

static std::map<std::string, char> binToHex = {
	{"0000", '0'},
	{"0001", '1'},
	{"0010", '2'},
	{"0011", '3'},
	{"0100", '4'},
	{"0101", '5'},
	{"0110", '6'},
	{"0111", '7'},
	{"1000", '8'},
	{"1001", '9'},
	{"1010", 'a'},
	{"1011", 'b'},
	{"1100", 'c'},
	{"1101", 'd'},
	{"1110", 'e'},
	{"1111", 'f'},
};

std::string toHex(std::string binary) {
	std::string hex = "0x";
	for (int i = 0; i <= 8; i += 4) {
		hex += binToHex[binary.substr(i, 4)];
	}
	return hex;
}

template <typename Iter>
std::string join(Iter begin, Iter end, std::string const& sep) {
	std::ostringstream result;
	if (begin != end)
		result << *begin++;
	while (begin != end)
		result << sep << *begin++;

	return result.str();
}